File: backend/packages/users/src/services/get.py
import redis
import psycopg2.extras
from psycopg2.extensions import connection
from ..utils.service import BaseService


class GetService(BaseService):
    _redis_client: redis.Redis
    _postgres_client: connection

    def __init__(self, redis_client: redis.Redis, postgres_client: connection) -> None:
        self._redis_client = redis_client
        self._postgres_client = postgres_client

    def call(self, user_id: str):
        user = None

        with self._postgres_client.cursor(
            cursor_factory=psycopg2.extras.RealDictCursor
        ) as cursor:
            cursor.execute(
                """
                SELECT u.user_id, u.email, u.name
                FROM users u
                WHERE u.user_id = %s
                """,
                (user_id,),
            )
            user = cursor.fetchone()

        if user is None:
            return {"reason": "Not found"}

        return {"user": user}

File: backend/packages/tokens/src/services/get.ts

import redis from 'redis'
import pg from 'pg'
import { BaseService } from '../utils/service'

type Input = {
    connectionId: string
    tokenId: string
    projectId: string
}

interface GetService extends BaseService<Input, any> { }

type Self = {
    redisClient: redis.RedisClientType
    postgresClient: pg.Pool
}

export const makeGetService = (redisClient: redis.RedisClientType, postgresClient: pg.Pool): GetService => {
    const self: Self = {
        redisClient,
        postgresClient
    }

    return {
        call: call(self)
    }
}

const call = (self: Self): GetService['call'] => async (input) => {
    const connectionId = input.connectionId
    const tokenId = input.tokenId
    const projectId = input.projectId

    const rawConnection = await self.redisClient.HGET("connections", connectionId)
    const connection = rawConnection && JSON.parse(rawConnection)
    const user = connection?.user

    if (!user) {
        return {
            reason: 'Not authenticated'
        }
    }

    const userId = user.user_id

    const projects = await self.postgresClient.query(`
        SELECT *
        FROM projects p
        WHERE p.project_id = $1 AND p.user_id = $2
    `, [projectId, userId])

    if (projects.rows.length === 0) {
        return {
            reason: 'Not authorized'
        }
    }

    const tokens = await self.postgresClient.query(`
        SELECT *
        FROM tokens t
        WHERE t.token_id = $1 AND t.project_id = $2
    `, [tokenId, projectId])

    return {
        token: tokens?.rows?.[0] || null
    }
}